TypeScript ESLint

dimpar opened the pull request:
In this PR we are adding a TypeScript linting config that can be imported by any project under Thesis organization.

dimpar committed:
Adding basic setup

dimpar committed:
Adding .tsconfig-eslint.json and readme

mhluongo commented on LICENSE:
```diff
@@ -0,0 +1,674 @@
+                    GNU GENERAL PUBLIC LICENSE
```
Let's use MIT for this

dimpar replied:
Done in cbd5e2b

mhluongo commented on package.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "eslint-config-typescript",
```
We should use a scoped name and make this just `eslint-config`, as TS is a superset of JS and that name can be auto imported. On NPM, that's `@thesis-co/eslint-config`. I can follow up with a PR to do builds + releases

mhluongo replied:
If you *do* want a JS-only config and to keep this at `eslint-config-typescript`, it would be good to see what that looks like and extend it here... maybe that was the intent?

dimpar replied:
I wanted to have a TS-only config here. I've changed it as you suggested in your first comment cbd5e2b Good point on scoping the name.

nkuba replied:
In my other comment, I suggested moving https://github.com/keep-network/eslint-config-keep under `thesis` org. In such a case, we need to make sure the names don't clash.

> If you do want a JS-only config and to keep this at eslint-config-typescript, it would be good to see what that looks like and extend it here... maybe that was the intent?

I like this suggestion very much (`@thesis-co/eslint-config-typescript` extends `@thesis-co/eslint-config`), it could be worth checking if it's possible.

mhluongo commented on package.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "eslint-config-typescript",
+  "version": "1.0.0",
+  "description": "ESLint typescript sharable config for thesis wide projects",
```
Capitalize "typescript" and "thesis", and s/sharable/shareable/

dimpar replied:
Done in 16e1f0c

mhluongo commented on package.json:
```diff
@@ -0,0 +1,22 @@
+  "description": "ESLint typescript sharable config for thesis wide projects",
+  "main": "index.js",
+  "repository": "https://github.com/thesis/eslint-config-typescript.git",
+  "author": "thesis",
```
Ideally you'd use your name + thesis.co email here

dimpar replied:
Done in 16e1f0c

mhluongo commented on README.md:
````diff
@@ -1 +1,44 @@
+ 2. In your `.eslintrc` add:
+ ```js
+{
+  "extends": ["eslint-config-typescript"],
````
If we rename the package, this can be `["@thesis-co"]` with the `eslint-config` assumed

dimpar replied:
I wasn't aware we can do this, done in 5cefeed

mhluongo reviewed (changes requested):
Could small changes. Once they land I can merge and fast-follow with an NPM release.

mhluongo commented on index.js:
```diff
@@ -0,0 +1,15 @@
+
+module.exports = {
+  extends: [
+    "eslint:recommended",
```
Rationale for skipping `"airbnb", "airbnb-typescript"` in this list?

dimpar replied:
Here db111be I've added `airbnb` support for typescript linting. I didn't notice `airbnb-typescript` at first and thought that in terms of `.js` files we could run into conflicts, cause at least in some `keep-*` projects we use `google` standard. Maybe it's a good idea to standardize linting across all `Thesis` projects.

dimpar committed:
Replacing GNU license with MIT

dimpar committed:
Fixing grammar in description and adding author email

dimpar committed:
Using scoped name @thesis-co/eslint-config

"eslint-config" can be auto imported when having @thesis-co/eslint-config
on NPM

dimpar committed:
Extending just @thesis-co with the eslint-config assumed

dimpar committed:
Using airbnb typescript linting for coding best practices

nkuba commented on index.js:
```diff
@@ -0,0 +1,16 @@
+"use strict"
+
+module.exports = {
```
We should make eslint configurations for TS and JS consistent. We already have JS configuration under the keep-network org: https://github.com/keep-network/eslint-config-keep. 

We probably need to move `@keep-network/eslint-config-keep` to become `@thesis-co/eslint-config`.

Right now we have:
- `@keep-network/eslint-config-keep` extends `google`,
- `@thesis-co/eslint-config-typescript` extends `airbnb`.

I don't have strong preferences between `airbnb` or `google`, but we should decide on one, unless we intentionally want to use them the way we do right now.

Maybe we could make it even more elegant by implementing what @mhluongo suggested in https://github.com/thesis/eslint-config-typescript/pull/1#discussion_r717558444 (`@thesis-co/eslint-config-typescript` extends `@thesis-co/eslint-config`).

dimpar replied:
I also don't have a strong preference regarding what linting standards we use. The initial commit by Matt was with `airbnb` and we can leave it. After having this discussion, I think we're going to import this repo only for new projects. "Older" keep-* projects will still use the existing `@keep-network/eslint-config-keep` config.

Regarding extending `@thesis-co/eslint-config` I'll see what can be done. `airbnb-typescript` needs `airbnb` to work. If we can avoid duplication, then yea, we can probably extend it.

nkuba commented on index.js:
```diff
@@ -0,0 +1,16 @@
+    "plugin:@typescript-eslint/recommended-requiring-type-checking"
+  ],
+  parser: "@typescript-eslint/parser",
+  plugins: ["@typescript-eslint"],
```
In [keep-network/eslint-config-keep](https://github.com/keep-network/eslint-config-keep/blob/0c27ade54e725f980e971c3d91ea88bab76b2330/index.js#L19-L22) we use `prettier`, do we also want to have it here?

dimpar replied:
I see it's growing :) the intent was to lint typescript only, but since we're heading towards two repos `@thesis-co/eslint-config` and `@thesis-co/eslint-config-typescript` then I can add `prettier` to `@thesis-co/eslint-config`

nkuba commented on package.json:
```diff
@@ -0,0 +1,24 @@
+  "main": "index.js",
+  "repository": "https://github.com/thesis/eslint-config-typescript.git",
+  "author": "dymitr.paremski@thesis.co",
+  "license": "GPL-3.0",
```
`LICENSE` file declares `MIT`, let's be consistent.

mhluongo reviewed (approved):
Looks good to me.
